File: engine/src/ecs/collider.c
/*
    This file is a part of yoyoengine. (https://github.com/zoogies/yoyoengine)
    Copyright (C) 2023  Ryan Zmuda

    Licensed under the MIT license. See LICENSE file in the project root for details.
*/

#include <yoyoengine/utils.h>
#include <yoyoengine/ecs/ecs.h>
#include <yoyoengine/ecs/collider.h>

/////////////////////////
//         v2          //
/////////////////////////

void ye_add_static_rect_collider_component(struct ye_entity *entity, float x, float y, float w, float h){
    struct ye_component_collider *collider = malloc(sizeof(struct ye_component_collider));
    collider->active = true;

    collider->is_trigger = false;
    collider->type = YE_COLLIDER_RECT;
    collider->x = x;
    collider->y = y;
    collider->width = w;
    collider->height = h;

    entity->collider = collider;
    ye_entity_list_add(&collider_list_head, entity);
}

void ye_add_trigger_rect_collider_component(struct ye_entity *entity, float x, float y, float w, float h){
    ye_add_static_rect_collider_component(entity, x, y, w, h);
    entity->collider->is_trigger = true;
}

void ye_add_static_circle_collider_component(struct ye_entity *entity, float x, float y, float radius){
    struct ye_component_collider *collider = malloc(sizeof(struct ye_component_collider));
    collider->active = true;

    collider->is_trigger = false;
    collider->type = YE_COLLIDER_CIRCLE;
    collider->x = x;
    collider->y = y;
    collider->radius = radius;

    entity->collider = collider;
    ye_entity_list_add(&collider_list_head, entity);
}

void ye_add_trigger_circle_collider_component(struct ye_entity *entity, float x, float y, float radius){
    ye_add_static_circle_collider_component(entity, x, y, radius);
    entity->collider->is_trigger = true;
}

/////////////////////////

void ye_remove_collider_component(struct ye_entity *entity){
    free(entity->collider);
    entity->collider = NULL;
    ye_entity_list_remove(&collider_list_head, entity);
}

File: engine/src/lua_runtime/ecs/collider.lua
--[[
    This file is a part of yoyoengine. (https://github.com/zoogies/yoyoengine)
    Copyright (C) 2023-2025  Ryan Zmuda

    Licensed under the MIT license. See LICENSE file in the project root for details.
]]

---@class Collider
---@field isActive boolean Controls whether the camera is active
---@field isRelative boolean Controls whether the position of the camera is relative to a root transform
---@field x number The x position
---@field y number The y position
---@field w number The width
---@field h number The height
---@field isTrigger boolean Whether the collider is a trigger
Collider = {
    -- no **real** fields.
    -- This exists purely for intellisense
}

local colliderIndexer = {
    isActive = 1,
    isRelative = 2,
    x = 3,
    y = 4,
    w = 5,
    h = 6,
    isTrigger = 7,
}

-- define the collider metatable
Collider_mt = {
    __index = function(self, key)
        return ValidateAndQuery(self, key, colliderIndexer, ye_lua_collider_query, "Collider")
    end,

    __newindex = function(self, key, value)
        return ValidateAndModify(self, key, value, colliderIndexer, ye_lua_collider_modify, "Collider")
    end,
}

---**Create a new collider component.**
---
---@param isTrigger boolean Whether the collider is a trigger
---@param x number The x position of the collider
---@param y number The y position of the collider
---@param w number The width of the collider
---@param h number The height of the collider
function Entity:AddColliderComponent(isTrigger, x, y, w, h) end -- fake prototype for intellisense
function AddColliderComponent(self, isTrigger,  x, y, w, h)
    if isTrigger then
        Entity:addComponent(self, ye_lua_create_trigger_collider, x, y, w, h)
    else
        Entity:addComponent(self, ye_lua_create_static_collider, x, y, w, h)
    end
end

---**Remove the collider component from the entity.**
function Entity:RemoveColliderComponent() end -- fake prototype for intellisense
function RemoveColliderComponent(self)
    ye_lua_remove_component(rawget(self, "_c_entity"), COLLIDER_COMPONENT)
end